src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -868,67 +616,82 @@ void CodegenDataModelProvider::InitDataModelForTesting()
     InitDataModelHandler();
 }
 
-std::optional<DataModel::DeviceTypeEntry> CodegenDataModelProvider::FirstDeviceType(EndpointId endpoint)
+DataModel::MetadataList<DataModel::DeviceTypeEntry> CodegenDataModelProvider::DeviceTypes(EndpointId endpointId)
 {
-    // Use the `Index` version even though `emberAfDeviceTypeListFromEndpoint` would work because
-    // index finding is cached in TryFindEndpointIndex and this avoids an extra `emberAfIndexFromEndpoint`
-    // during `Next` loops. This avoids O(n^2) on number of indexes when iterating over all device types.
-    //
-    // Not actually needed for `First`, however this makes First and Next consistent.
-    std::optional<unsigned> endpoint_index = TryFindEndpointIndex(endpoint);
+    std::optional<unsigned> endpoint_index = TryFindEndpointIndex(endpointId);
     if (!endpoint_index.has_value())
     {
-        return std::nullopt;
+        return {};
     }
 
     CHIP_ERROR err                            = CHIP_NO_ERROR;
     Span<const EmberAfDeviceType> deviceTypes = emberAfDeviceTypeListFromEndpointIndex(*endpoint_index, err);
-    SpanSearchValue<chip::Span<const EmberAfDeviceType>> searchable(&deviceTypes);
-
-    return DeviceTypeEntryFromEmber(searchable.First<ByDeviceType>(mDeviceTypeIterationHint).Value());
-}
 
-std::optional<DataModel::DeviceTypeEntry> CodegenDataModelProvider::NextDeviceType(EndpointId endpoint,
-                                                                                   const DataModel::DeviceTypeEntry & previous)
-{
-    // Use the `Index` version even though `emberAfDeviceTypeListFromEndpoint` would work because
-    // index finding is cached in TryFindEndpointIndex and this avoids an extra `emberAfIndexFromEndpoint`
-    // during `Next` loops. This avoids O(n^2) on number of indexes when iterating over all device types.
-    std::optional<unsigned> endpoint_index = TryFindEndpointIndex(endpoint);
-    if (!endpoint_index.has_value())
+    DataModel::MetadataList<DataModel::DeviceTypeEntry> result;
+    err = result.reserve(deviceTypes.size());
+    if (err != CHIP_NO_ERROR)
     {
-        return std::nullopt;
+#if CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
+        ChipLogError(AppServer, "Failed to reserve device type buffer space: %" CHIP_ERROR_FORMAT, err.Format());
+#endif
+        return {};
+    }
+    for (auto & entry : deviceTypes)
+    {
+        err = result.Append(DeviceTypeEntryFromEmber(entry));
```

Comment:
It sure would be nice if we just used the same data type and could just return the span we already have directly... Followup, please.

Reply:
We have an open issue for this. The issue is that we do not want to leak ember types out of data model provider, so we have to make ember use the provider type instead.

Reply:
Issue is https://github.com/project-chip/connectedhomeip/issues/35889. Closing this comment and we can update on that issue as a followup eventually.